MobileCenter/MobileCenter/Internals/Channel/MSChannelDefault.m

```diff
@@ -392,4 +405,114 @@ - (void)notifyFailureBeforeSendingForItem:(id<MSLog>)item withError:(NSError *)e
+          UIApplication *sharedApplication = [MSUtility sharedApplication];
+          if (sharedApplication && (self.backgroundTaskIdentifier != UIBackgroundTaskInvalid)) {
+            [sharedApplication endBackgroundTask:self.backgroundTaskIdentifier];
+            self.backgroundTaskIdentifier = UIBackgroundTaskInvalid;
+            MSLogDebug([MSMobileCenter logTag], @"No more logs to flush, invalidating the background task.");
+          }
+        }
+      }
+    });
+  }
+#endif
+}
+
+#pragma mark – Observers
+
+- (void)addObservers {
+
+// There is no need to do trigger sending on macOS because we can just continue to execute tasks there.
+#if !TARGET_OS_OSX
+  if (!MS_IS_APP_EXTENSION) {
+    self.backgroundTaskIdentifier = UIBackgroundTaskInvalid;
+    __weak typeof(self) weakSelf = self;
+    if (self.appDidEnterBackgroundObserver == nil) {
+      void (^notificationBlock)(NSNotification *note) = ^(NSNotification __unused *note) {
+        typeof(self) strongSelf = weakSelf;
+        strongSelf.isInBackground = YES;
+        if (self.timerSource != nil) {
+
+          /**
+           * From the documentation for applicationDidEnterBackground:
+           * "It's likely any background tasks you start in applicationDidEnterBackground: will not run until after that
+           * method exits, you should request additional background execution time before starting those tasks.
+           * In other words, first call beginBackgroundTaskWithExpirationHandler: and then run the task on a
+           * dispatch queue or secondary thread.
+           */
+          UIApplication *sharedApplication = [MSUtility sharedApplication];
+
+          // Checking if sharedApplication is != nil as it can be nil for extensions.
+          if (sharedApplication && self.backgroundTaskIdentifier == UIBackgroundTaskInvalid) {
+            self.backgroundTaskIdentifier = [sharedApplication beginBackgroundTaskWithExpirationHandler:^{
+              MSLogDebug([MSMobileCenter logTag], @"Background task has expired, invalidating it and suspend sender.");
+              [sharedApplication endBackgroundTask:self.backgroundTaskIdentifier];
+              self.backgroundTaskIdentifier = UIBackgroundTaskInvalid;
+
+              // Suspend the sender, don't suspend the channel itself.
+              [strongSelf.sender suspend];
+            }];
+          }
+
+          // We've requested additional background time and are now flushing our queue.
+          [strongSelf flushQueue];
+        }
+      };
+      self.appDidEnterBackgroundObserver =
+          [MS_NOTIFICATION_CENTER addObserverForName:UIApplicationDidEnterBackgroundNotification
+                                              object:nil
+                                               queue:NSOperationQueue.mainQueue
+                                          usingBlock:notificationBlock];
+    }
+
+    self.appWillEnterForegroundObserver =
```

Comment:
One last thing, you should invalidate the bg task here every time the app gets to the foreground if not already invalidated.

Reply:
The last commit changes `stopBackgroundActivity` so it will cancel the background activity every time the app is in the foreground, no matter if we have ongoing sender tasks. They will continue as the app is now in the foreground.